flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/fetch/PostgresScanFetchTask.java

```diff
@@ -0,0 +1,395 @@
+            PostgresSourceFetchTaskContext ctx,
+            SnapshotSplitChangeEventSourceContext changeEventSourceContext)
+            throws InterruptedException {
+        final StreamSplit backfillSplit =
+                new StreamSplit(
+                        split.splitId(),
+                        changeEventSourceContext.getLowWatermark(),
+                        changeEventSourceContext.getHighWatermark(),
+                        new ArrayList<>(),
+                        split.getTableSchemas(),
+                        0);
+
+        // optimization that skip the WAL read when the low watermark >=  high watermark
+        final boolean backfillRequired =
+                backfillSplit.getEndingOffset().isAfter(backfillSplit.getStartingOffset());
+        if (!backfillRequired) {
+            LOG.info(
+                    "Skip the backfill {} for split {}: low watermark >= high watermark",
+                    backfillSplit,
+                    split);
+            ctx.getDispatcher()
+                    .dispatchWatermarkEvent(
+                            ctx.getOffsetContext().getPartition(),
+                            backfillSplit,
+                            backfillSplit.getEndingOffset(),
+                            WatermarkKind.END);
+
+            taskRunning = false;
+            return;
+        }
+
+        final PostgresOffsetContext.Loader loader =
+                new PostgresOffsetContext.Loader(ctx.getDbzConnectorConfig());
+        final PostgresOffsetContext postgresOffsetContext =
+                loader.load(backfillSplit.getStartingOffset().getOffset());
+
+        // we should only capture events for the current table,
+        // otherwise, we may not find corresponding schema
+        Configuration dbzConf =
+                ctx.getDbzConnectorConfig()
+                        .getConfig()
+                        .edit()
+                        .with("table.include.list", split.getTableId().toString())
+                        // Disable heartbeat event in snapshot split fetcher
+                        .with(Heartbeat.HEARTBEAT_INTERVAL, 0)
+                        .build();
+
+        final PostgresStreamFetchTask.StreamSplitReadTask backfillReadTask =
+                new PostgresStreamFetchTask.StreamSplitReadTask(
+                        new PostgresConnectorConfig(dbzConf),
+                        ctx.getSnapShotter(),
+                        ctx.getConnection(),
+                        ctx.getDispatcher(),
+                        ctx.getErrorHandler(),
+                        ctx.getTaskContext().getClock(),
+                        ctx.getDatabaseSchema(),
+                        ctx.getTaskContext(),
+                        ctx.getReplicationConnection(),
+                        backfillSplit);
+        LOG.info("Execute backfillReadTask for split {}", split);
+        backfillReadTask.execute(new PostgresChangeEventSourceContext(), postgresOffsetContext);
```

Comment:
Could you elaborate on what you mean here?

Reply:
I see we could get the offsetContext from `ctx `. Why do we need to reload and use a new offsetContext. I do not get the point.

Reply:
I see. This is similar to the [MySQL experimental connector](https://github.com/goldsky-io/goldsky-flink-cdc-connectors/blob/master/flink-cdc-base/src/test/java/com/ververica/cdc/connectors/base/experimental/fetch/MySqlScanFetchTask.java#L154-L158). 

We need to set the LSN to the low watermark when backfilling, otherwise it always tries to start with LSN 0/0.
In fact, we were using `ctx. getOffsetContext()` initially, and then found it causes issues:)